Add Couchbase cache plugin

MikeGoldsmith opened the pull request:
- based on Couchbase Java SDK 2.7.4

daschl commented on parent/pom.xml:
```diff
@@ -85,6 +85,8 @@
       <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
       <plugins.chm/>
+      <plugins.couchbase/>
+      <plugins.couchbase-latest/>
```
why is this needed?

MikeGoldsmith replied:
I believe this is used to discover the plugin projects and is used by the build scripts when creating the distributable ZIP.

alanfx replied:
First of all, thank you for this contribution. This property allows you to build a RadarGun distribution that only includes the latest version of each plugin, instead of every version of the plugin.

daschl commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseService.java:
```diff
@@ -0,0 +1,45 @@
+
+   @Override
+   public void start() {
+      cluster = CouchbaseCluster.create(nodes);
```
I think since this is just doing sync kv ops, we can pass in the environment and set the following:

.callbacksOnIoPool(true)
.runtimeMetricsCollectorConfig(DefaultMetricsCollectorConfig.disabled())
.networkLatencyMetricsCollectorConfig(DefaultLatencyMetricsCollectorConfig.disabled())

daschl commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java:
```diff
@@ -0,0 +1,70 @@
+
+      @Override
+      public void clear() {
+         throw new UnsupportedOperationException("Clearing Couchbase is not implemented yet.");
```
does this compare to the flush() command on the bucket manager?

MikeGoldsmith replied:
Yes, it does - I didn't implement as the Redis plugin provider I used as a basis didn't implement it. I'm not sure if it's used but can update to do the flush.

daschl commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java:
```diff
@@ -0,0 +1,70 @@
+
+      @Override
+      public byte[] getAndPut(byte[] key, byte[] value) {
+         byte[] document = get(key);
```
I think this should use the cas on put (replace) to be more realistic.

Might make sense to have specialized impls for this to avoid creating a string key more than once

MikeGoldsmith replied:
Sure, will look to do that.

daschl commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java:
```diff
@@ -0,0 +1,70 @@
+
+      @Override
+      public byte[] getAndRemove(byte[] key) {
+         byte[] doc = get(key);
```
I think this should use the cas on remove to be more realistic.

Might make sense to have specialized impls for this to avoid creating a string key more than once

MikeGoldsmith replied:
Sure, will look to do that.

MikeGoldsmith commented:
@alanfx  @diegolovison I think this is ready for initial review. Let me know if there is anything else you would like to see in this PR.

diegolovison commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java:
```diff
@@ -0,0 +1,92 @@
+      return adapter;
+   }
+
+   private class CouchbaseCacheAdapter<K, V> implements BasicOperations.Cache<byte[], byte[]> {
```
You should use something like: `private class CouchbaseCacheAdapter <K, V> implements BasicOperations.Cache<K, V>`. See as an example: https://github.com/radargun/radargun/commit/373b353317d0a05b9a7e9e7009a3c376500b8d24#diff-df93a8f6e3d2569df9283e049b55bfbc

RadarGun allows different keys and values generators. Here are some that we support https://github.com/radargun/radargun/tree/master/extensions/cache/src/main/java/org/radargun/stages/cache/generators

diegolovison commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java:
```diff
@@ -0,0 +1,92 @@
+
+import java.nio.charset.StandardCharsets;
+
+public class CouchbaseOperations implements BasicOperations {
```
I prefer to see the name of the class like `CouchbaseClientOperations`

diegolovison commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseService.java:
```diff
@@ -0,0 +1,56 @@
+import org.radargun.traits.Lifecycle;
+
+@Service(doc = "Couchbase")
+public class CouchbaseService implements Lifecycle {
```
Same here: CouchbaseClientService

diegolovison commented on plugins/couchbase/src/main/resources/plugin.properties:
```diff
@@ -0,0 +1,3 @@
+service.default org.radargun.service.CouchbaseService
```
For me, it is missing the `CouchbaseServerService` class implementation. For example, I would like to run RadarGun and compare the project X with Couchbase in a perf lab. Without `CouchbaseServerService` I will need to setup the server everytime. I saw https://docs.couchbase.com/server/4.1/getting-started/installing.html that we can download and install the server. A good entry point is https://github.com/radargun/radargun/blob/master/plugins/etcd/src/main/java/org/radargun/service/EtcdServerService.java

diegolovison commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseOperations.java:
```diff
@@ -0,0 +1,92 @@
+      public byte[] getAndPut(byte[] keyBytes, byte[] value) {
+         String key = createKey(keyBytes);
+         byte[] document = get(key);
+         put(key, value);
```
@rvansa should it be in a unique operation or doesn't support it, right?

MikeGoldsmith commented:
@diegolovison I've updated the class names / files and refactored `CouchbaseClientOperations` to utilize **K** and **V** type params in the inner `CouchbaeCacheAdapter`.

I'm not sure downloading and installing Couchbase as part of a test run is a good idea. It is a fairly large process / download. There are docker images and/or vagrant scripts that we may be able to utilize.

diegolovison commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientOperations.java:
```diff
@@ -0,0 +1,105 @@
+      @Override
+      public V getAndRemove(K key) {
+         String keyStr = keyAsString(key);
+         V document = get(keyStr);
```
It should be only one operation in the server. In this case, we are calling the server twice. If Couchbase doesn't support returning the previous value when removing the key we need to throw an exception.

diegolovison commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientOperations.java:
```diff
@@ -0,0 +1,105 @@
+      public V getAndPut(K key, V value) {
+         String keyStr = keyAsString(key);
+         V document = get(keyStr);
+         put(keyStr, value);
```
It should be only one operation in the server. In this case, we are calling the server twice. If Couchbase doesn't support returning the previous value when putting the key we need to throw an exception.

diegolovison commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientOperations.java:
```diff
@@ -0,0 +1,105 @@
+package org.radargun.service;
```
I just saw that the folder structure is using (dot) instead of a folder separator.
The name of the file is: `org.radargun.service/CouchbaseClientOperations.java` and it should be `org/radargun/service/CouchbaseClientOperations.java`

diegolovison commented on plugins/couchbase/src/main/java/org.radargun.service/CouchbaseClientService.java:
```diff
@@ -0,0 +1,56 @@
+package org.radargun.service;
```
Same here about the folder separator

diegolovison commented:
@MikeGoldsmith I am going to squash and merge using the GitHub button. IIRC, the commit will be in my name. I would like to see the commit with who developed that. Would you like to squash or can I move forward?

diegolovison commented:
@MikeGoldsmith I was going to use the "Squash and merge" button in the GitHub interface and I just remembered that last time that I did it, I was the commit author. I prefer to see your team as an author of this commit. Could you please squash?

MikeGoldsmith commented:
@diegolovison commits have been squashed

MikeGoldsmith commented:
@diegolovison commits have been squashed 👍

diegolovison commented:
Thanks @MikeGoldsmith and @daschl
